flink-runtime/src/main/java/org/apache/flink/runtime/rpc/akka/AkkaRpcServiceBuilder.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.rpc.akka;
+
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.runtime.akka.AkkaUtils;
+import org.apache.flink.runtime.clusterframework.BootstrapTools;
+import org.apache.flink.util.NetUtils;
+import org.apache.flink.util.Preconditions;
+
+import akka.actor.ActorSystem;
+import com.typesafe.config.Config;
+import org.slf4j.Logger;
+
+import javax.annotation.Nullable;
+
+import java.net.InetAddress;
+
+/**
+ * Builder for {@link AkkaRpcService}.
+ */
+public class AkkaRpcServiceBuilder {
```

Comment:
I don't get it.
- Constructor of `AkkaRpcService` is used in various of test cases. Unless we want to modify them all, I don't think we can make the constructor private. Maybe it makes sense to annotate the constructor as `VisibleForTesting`, because without the test cases it can be package visibility.
- If we cannot make constructor of `AkkaRpcService` private anyway, then I would prefer to put `AkkaRpcServiceBuilder` in a separate file. The file `AkkaRpcService.java` is already quite long. No need to further complicate it.
- I don't see how `AkkaRpcService#local(remote)RpcServiceBuilder` is more readable compared to `AkkaRpcServiceUtils#local(remote)ServiceBuilder`.
- It probably make sense to make the constructor of the builder private, since we already have `AkkaRpcServiceUtils#local(remote)ServiceBuilder`. In that case, I would tend to put the builder inside `AkkaRpcServiceUtils` rather than `AkkaRpcService`.

Reply:
I think whether we make it an inner class or not should only depend on whether we need to leverage the characteristic of the inner class mechanism. It has nothing to do with how many tests we need to modify or how complicated the `AkkaRpcService` is.

In this case, we could get benefit from making it an inner class of `AkkaRpcService` is:
- For both AkkaRpcService and Builder, we could prevent unexpected access to their constructor.
- We could expose two static functions whose use case is well explained by the method name, instead of two constructors only different in param list.

The alternative, which annotates the constructor as `VisibleForTesting` and put the builder inside `AkkaRpcServiceUtils`, also LGTM. In this way, we could only build an `AkkaRpcService` through the utility class. If it needs lots of effort to change the test cases, I think it would be good.

Reply:
I think whether we want to make a constructor private, depends on whether allowing it to be accessed from outside may cause problem. E.g., the none argument constructor of `ResourceProfile` must be private because the only case we need it is for constructing the singleton instance `UNKNOWN`. If we allow to create another instance of `ResourceProfile` with this none argument constructor, the created instance may not work properly.

In this case, I don't see such problem in allowing direct access to the constructors. That makes hiding the constructors a nice but not must to have approach, by which we suggest people to choose the more convenient builders. 

On the other hand, it is not about how many efforts needed that we do not use builders in the test cases. Currently, the builder does two things: create a `ActorSystem`, and use it to create a `AkkaRpcService`. However, in many test cases, we need to use a custom testing actor system, which is different from the one we created for production. That's way we cannot easily use builders in the tests.

We could also make the builder to support custom actor system. But that complicates the builder logics (what should we do if both custom actor system and the fields for creating the production actor system are provided), and I think it doesn't make sense to do for pure testing purpose.

Reply:
I moved `AkkaRpcServiceBuilder` into `AkkaRpcServiceUtils`, and made its constructors private. Also annotated constructor of `AkkaRpcService` `VisibleForTesting`.